Extend `cilium config` to expose all active configurations. Add subcommand `cilium config get` to get configurations from CLI

h3llix opened the pull request:
Fixes: #13913 

For `cilium config get <config_name>` configuration name should be in kebab case. For example `EncryptNode`, So we can get its value by `cilium config get encrypt-node`. Add flag -r to display read-only configurations. 

Test-image : `docker.io/h3llix/cilium-dev:cilium-config` (if required)

![image](https://user-images.githubusercontent.com/55123900/127231087-7128e308-62ea-410e-a917-a3179de3c188.png)
![image](https://user-images.githubusercontent.com/55123900/127231252-6c2f91c7-2170-44b8-a4ca-261a010c8ff5.png)
![image](https://user-images.githubusercontent.com/55123900/127231287-92e38f39-31e7-4dce-b5f1-aa286e6ca778.png)

h3llix committed:
api,cli: expose active cilium config through CLI

Note: All the daemon configurations have to be public for reflections to work

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

aanm reviewed (commented):
It looks it's going on the right track. I assume all fields had to be made public otherwise the reflect won't work?

pchaigno commented on pkg/option/config.go:
```diff
@@ -1493,6 +1494,7 @@ type DaemonConfig struct {
 
 	// ToFQDNsMaxIPsPerHost defines the maximum number of IPs to retain for
 	// expired DNS lookups with still-active connections
+
```
Looks like this newline was added by mistake.

pchaigno commented on pkg/option/config.go:
```diff
@@ -1456,11 +1456,12 @@ type DaemonConfig struct {
+	EnableBPFMasquerade  bool
+	EnableBPFClockProbe  bool
+	EnableIPMasqAgent    bool
+	// expired DNS lookups with still-active connections
```
What is this change?

h3llix replied:
I am not sure if I made any change here. But on rechecking it seems my editor formatted that.

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -75,6 +75,12 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+		// 	fmt.Println("key  ", k, "   value", v)
+		// }
+		for k, v := range cfgStatus.DaemonConfigurationMap {
+			fmt.Println(k, " ", v)
```
We should try and align the values as much as possible.

pchaigno reviewed (commented):
LGTM! A few minor comments below.

I've answered to your weekly report to discuss what we want to include in that PR before merging.

h3llix committed:
cli: add subcommand `cilium config get`

cilium config get <config_name> gives configuration value from the cli.
conifg_name should be in kebab case.

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

h3llix commented on pkg/option/config.go:
```diff
@@ -2121,7 +2123,7 @@ func (c *DaemonConfig) LocalClusterName() string {
 // handled.
 func (c *DaemonConfig) K8sServiceProxyName() string {
-	return c.k8sServiceProxyName
+	return c.K8sServiceProxyNametest
```
Need some suggestions here. Viper is also using the same flag name. Here: https://github.com/cilium/cilium/blob/d5ff6879dbc50de93cde07b4e6c87f2581106f34/pkg/option/config.go#L2496 And all the Configurations need to be public.

aanm replied:
Rename the function name with `K8sServiceProxyNameValue`

h3llix replied:
Yes sure, Also do we really need these public functions to return config value from the struct? The reason I am asking this is that  we are already making the configurations public.

h3llix commented on api/v1/openapi.yaml:
```diff
@@ -2198,6 +2198,10 @@ definitions:
         description: Configured compatibility mode for --egress-multi-home-ip-rule-compat
         type: boolean
+      daemonConfigurationMap:
+        description: test
```
I will also fix the description here.

aanm commented on pkg/option/config.go:
```diff
@@ -1755,8 +1756,8 @@ type DaemonConfig struct {
-	// ipv4NativeRoutingCIDR describes a CIDR in which pod IPs are routable
-	ipv4NativeRoutingCIDR *cidr.CIDR
+	// Ipv4NativeRoutingCIDR describes a CIDR in which pod IPs are routable
+	Ipv4NativeRoutingCIDR *cidr.CIDR
```
nit, this could be renamed to `IPv4NativeRoutingCIDR` (instead of `Ipv4NativeRoutingCIDR`)

h3llix replied:
I used that variable name purposefully because there was a function with the same name which was being called with `DaemonConfig` struct. Now, I just changed the function name.

aanm commented on cilium/cmd/config.go:
```diff
@@ -73,18 +75,39 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+			if reflect.ValueOf(v).Kind() == reflect.Map {
+				mapString := make(map[string]string)
+				for key, value := range v.(map[string]interface{}) {
+					strKey := fmt.Sprintf("%v", key)
```
`key` is already a string so this line does not seem to be necessary.

aanm commented on cilium/cmd/config.go:
```diff
@@ -73,18 +75,39 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+			v := cfgStatus.DaemonConfigurationMap[k]
+			if reflect.ValueOf(v).Kind() == reflect.Map {
+				mapString := make(map[string]string)
+				for key, value := range v.(map[string]interface{}) {
```
What guarantees do we have that `v` is always capable of being cast to `map[string]interface{}`? Might be worth checking if we can do it before doing this for-loop with `map, ok := v.(map[string]interface{}); if ok { do for loop }`

aanm commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,53 @@
+// Copyright 2017-2018 Authors of Cilium
```
minor nit: `Copyright 2017-2018` -> `Copyright 2021`

h3llix committed:
cli: seperate read-only and read-write configuraitons

Add -r flag to show read-only configurations while default cilium config
show read-write configurations

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

aanm commented on cilium/cmd/config.go:
```diff
@@ -73,15 +75,40 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+					}
+					dumpConfig(mapString)
+					continue
+				}
```
if !ok we should print an warning / error stating that.

aanm reviewed (changes requested):
Just one minor thing.

joestringer commented on cilium/cmd/config.go:
```diff
@@ -73,15 +75,42 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+					dumpConfig(mapString)
+					continue
+				} else {
+					fmt.Print("Error: cannot cast daemon config map to map[string]interface{}\n")
```
To make sure that errors are separated from the output from the tool, please consider printing errors to stderr like this:

```suggestion
					fmt.Fprint(os.Stderr, "Error: cannot cast daemon config map to map[string]interface{}\n")
```

This can be helpful to ensure that the tool plays well with other unix tools (`grep`, etc.).

joestringer commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,53 @@
+		if id, ok := configMap[configName]; ok {
+			fmt.Printf("%v\n", id)
+		} else {
+			fmt.Printf("Configuration does not exist\n")
```
You might consider making this one `Fatalf()` as well. The user requested "please give me configuration `foo`", but if we can't satisfy that request then it probably makes sense to return an error return value on the CLI. This way, anything that calls this CLI can know whether the operation successfully retrieved the value, or failed to retrieve the value just based on the exit code.

joestringer commented on pkg/k8s/utils/utils.go:
```diff
@@ -72,9 +72,9 @@ func GetObjNamespaceName(obj NamespaceNameGetter) string {
 // ServiceConfiguration is the required configuration for GetServiceListOptionsModifier
 type ServiceConfiguration interface {
-	// K8sServiceProxyName must return the value of the proxy name
+	// K8sServiceProxyNameValue must return the value of the proxy name
```
nit: Why is this rename necessary? The new name seems more confusing than the current name.

h3llix replied:
Hey @joestringer. Thanks for the review! Actually, there was a field name `k8sServiceProxyName` which had to be made public for reflections to work. As well there was already a function named `K8sServiceProxyName` to get the value from the struct. So @aanm asked me to change the function name to `K8sServiceProxyNameValue`.

h3llix replied:
Maybe we can rename it to `GetK8sServiceProxyName`.

joestringer replied:
Aha, it was the reflection piece that made it necessary, I see. Reading through I couldn't tell what it conflicted with since the conflicting name isn't touched in this PR at all (maybe because it's already public?).

In this case, I mildly prefer `GetK8sServiceProxyName` but it doesn't make that big of a difference. Don't worry about it :)

joestringer commented on cilium/cmd/config.go:
```diff
@@ -73,15 +75,42 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 				os.Exit(1)
 			}
+			if err := command.PrintOutput(cfgStatus.DaemonConfigurationMap); err != nil {
+				os.Exit(1)
```
I realize that the existing code already did this, but we should really do better handling on these errors to print the problem:

```suggestion
				Fatalf(err)
```

joestringer commented on cilium/cmd/config.go:
```diff
@@ -73,15 +75,42 @@ func configDaemon(cmd *cobra.Command, opts []string) {
-		fmt.Printf("%-24s %s\n", "k8s-endpoint", cfgStatus.K8sEndpoint)
-		fmt.Printf("%-24s %s\n", "PolicyEnforcement", cfgStatus.Realized.PolicyEnforcement)
+		fmt.Printf("%-34s: %s\n", "k8s-configuration", cfgStatus.K8sConfiguration)
+		fmt.Printf("%-34s: %s\n", "k8s-endpoint", cfgStatus.K8sEndpoint)
```
I don't think that these `k8s-configuration`, `k8s-endpoint` options are read/write. In fact in my minikube they don't even print anything:

```
# cilium config 
Conntrack                Enabled
ConntrackAccounting      Enabled
ConntrackLocal           Disabled
Debug                    Disabled
DebugLB                  Disabled
DropNotification         Enabled
MonitorAggregationLevel  Medium
PolicyAuditMode          Disabled
PolicyTracing            Disabled
PolicyVerdictNotification Enabled
TraceNotification        Enabled
k8s-configuration        
k8s-endpoint             
PolicyEnforcement        default
```

We can at least put them in the read-only section, or if someone knows we don't need them, maybe even remove them.

h3llix replied:
Yes, I'll make the necessary changes. For now, I will put them in the read-only section.

h3llix replied:
@joestringer Can we have a small discussion with someone about the unnecessary fields in the cli output?

joestringer replied:
@h3llix I've started a thread here around it: https://cilium.slack.com/archives/C2B917YHE/p1626885955046300

Maybe it's just a bug that we used to print some k8s information for these two fields, but we no longer print it. For now I'm OK with just moving these to read-only, we can work on getting this PR into the tree. Later on if you or others are interested into digging more into why those fields are not printing anything, we can investigate more.

joestringer commented on cilium/cmd/config.go:
```diff
@@ -73,15 +75,42 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 		if cfgStatus.NodeMonitor != nil {
-			fmt.Printf("%-24s %d\n", "MonitorNumPages", cfgStatus.NodeMonitor.Npages)
+			fmt.Printf("%-34s: %d\n", "MonitorNumPages", cfgStatus.NodeMonitor.Npages)
 		}
```
I don't think that this `MonitorNumPages` option is read/write either.

joestringer reviewed (changes requested):
Thanks for working on this! It looks good overall, I just have a few minor nits here and there. The main one is that some of the options under the read-write section seem like read-only options to me. Then there's just some various general CLI output "best-practices" which I've made suggestions about.

nathanjsweet commented on cilium/cmd/config.go:
```diff
@@ -73,16 +75,43 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+					fmt.Fprintf(os.Stderr, "Error: cannot cast daemon config map to map[string]interface{}\n")
+				}
+			}
+			fmt.Printf("%-34s: %v\n", k, v)
```
If `ok` is `false` above then this line will run. That doesn't seem right. I would recommend putting this in an `else` statement.

h3llix replied:
Sorry for the delay. Thanks for pointing that out. I will put a fix.

nathanjsweet reviewed (changes requested):
Overall it looks good.

joestringer reviewed (changes requested):
I did a bit of brief testing, and I noticed that the default JSON output is broken:

```
# cilium config -o json
{
  "Conntrack": "Enabled",
  "ConntrackAccounting": "Enabled",
  "ConntrackLocal": "Disabled",
  "Debug": "Disabled",
  "DebugLB": "Disabled",
  "DropNotification": "Enabled",
  "MonitorAggregationLevel": "Medium",
  "PolicyAuditMode": "Disabled",
  "PolicyTracing": "Disabled",
  "PolicyVerdictNotification": "Enabled",
  "TraceNotification": "Enabled"
}

##### Read Write Configurations #####
Conntrack                         : Enabled
ConntrackAccounting               : Enabled
ConntrackLocal                    : Disabled
Debug                             : Disabled
DebugLB                           : Disabled
DropNotification                  : Enabled
MonitorAggregationLevel           : Medium
PolicyAuditMode                   : Disabled
PolicyTracing                     : Disabled
PolicyVerdictNotification         : Enabled
TraceNotification                 : Enabled
PolicyEnforcement                 : default
```

The `-o json` output must output only valid JSON.

pchaigno commented on pkg/option/config.go:
```diff
@@ -1448,11 +1448,12 @@ type DaemonConfig struct {
+	EnableBPFMasquerade  bool
+	EnableBPFClockProbe  bool
+	EnableIPMasqAgent    bool
+	// expired DNS lookups with still-active connections
```
This change still needs to be reverted.

h3llix replied:
I actually reverted this once in one of my previous commits. But I remember the tests were giving me lint issues with that.

pchaigno replied:
Can you try again and share the lint issues? Maybe those were on master and it was fixed since then.

h3llix replied:
@pchaigno I just added a test commit to check if it fails

christarazi replied:
It is still failing. I suggest running `contrib/scripts/check-fmt.sh` locally and adding `-w` to the `gofmt` command to write the result back to the original file. If that is still not working, your Golang version might be out of date and you should consider updating to the latest.

h3llix replied:
Hi @christarazi. Actually the test commit is intentional whose linting is as same as master. I had a conversation with @pchaigno and he wanted to confirm whether those lint issues were on my master or my commits.

pchaigno commented on cilium/cmd/helpers.go:
```diff
@@ -377,11 +388,43 @@ func dumpConfig(Opts map[string]string) {
+				fmt.Printf("\t%-26s: %s\n", k, value)
+			} else {
+				fmt.Printf("%-34s: %s\n", k, value)
+			}
```
The `if indented {...} else {...}` could be moved to line 388, to simplify the code here. Basically, you'd have only:
```go
formatStr := "%-34s: %s\n"
if indented {
    formatStr = "\t%-26s: %s\n"
}
```

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 		}
 	} else if len(opts) == 0 {
```
The contents of this else if block has gotten a lot more complex. I think it would be worth moving it to its own dedicated function.

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
-				os.Exit(1)
+			if listReadOnlyConfigurations {
+				if err := command.PrintOutput(cfgStatus.DaemonConfigurationMap); err != nil {
+					Fatalf("Error while printing configmap: %v", err)
```
```suggestion
					os.Exit(1)
```

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
-		if cfgStatus.NodeMonitor != nil {
-			fmt.Printf("%-24s %d\n", "MonitorNumPages", cfgStatus.NodeMonitor.Npages)
+		if listReadOnlyConfigurations {
+			fmt.Print("### Read-Only Configurations ###\n\n")
```
```suggestion
			fmt.Print("### Read-Only Configurations ###\n")
```

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+			}
+			dumpConfig(cfgStatus.Immutable, false)
+		} else {
+			fmt.Print("##### Read Write Configurations #####\n")
```
Nit:
```suggestion
			fmt.Print("##### Read-Write Configurations #####\n")
```

pchaigno commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+			fmt.Printf("%-34s: %s\n", "k8s-configuration", cfgStatus.K8sConfiguration)
+			fmt.Printf("%-34s: %s\n", "k8s-endpoint", cfgStatus.K8sEndpoint)
+			if cfgStatus.NodeMonitor != nil {
+				fmt.Printf("%-34s: %d\n", "MonitorNumPages", cfgStatus.NodeMonitor.Npages)
```
Why was this option made read-only? (It's the size of the perf ring buffer we use to send notifications to the agent.)

joestringer commented on cilium/cmd/config.go:
```diff
@@ -70,20 +76,55 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+			fmt.Printf("%-34s: %s\n", "PolicyEnforcement", cfgStatus.Realized.PolicyEnforcement)
+			return
 		}
-		return
```
nit: It's slightly odd to remove the return from after the if/else check and only add the `return` inside one of the `if` / `else` statements, maybe the return should just remain at the end like it was before?

There's a similar thing happening on lines 83, 88. Both sides of the conditional statement execute `return`, so there could just be one `return` statement after the `if`/`else` rather than two.

Functionally there's no difference though and this code is not likely to break in future because the `return`s are in unusual places so I'm OK to just let this be.

pchaigno commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+var (
+	re = regexp.MustCompile("[^0-9a-zA-Z]")
```
Wouldn't `[^\w]` work here?

Could you also rename the variable to better clarify its purpose?

pchaigno commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,69 @@
+		if value, ok := configMap[configName]; ok {
+			fmt.Printf("%v\n", value)
+		} else {
+			Fatalf("Configuration does not exist\n")
```
```suggestion
			Fatalf("Configuration does not exist")
```

joestringer reviewed (changes requested):
Now that there's this new option `-r`, I'd kinda like to see an `--all` flag so we can ensure to grab everything in one command. :thinking: Maybe we can work on something like that in a separate follow-up.

There's one more request I have, to add the new command (using `-r` flag) to bugtool/ (since it's separate from the existing `cilium config`). This will ensure that when users report bugs, we can gather this information with the sysdump. You can find the right place for the command here:

```
$ git grep "cilium config" bugtool/
bugtool/cmd/configuration.go:           "cilium config",
```

pchaigno commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,69 @@
+
+		configMap := mergeMaps(readOnlyConfigMap, readWriteConfigMap)
+
+		if value, ok := configMap[configName]; ok {
```
Could you add a small comment to explain what is being done on lines 36 and 53 to 58? We've talked about it, so it's clear to me, but it may come as a bit strange to others.

pchaigno commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,69 @@
+	Short:  "Retrieve cilium configuration",
+	PreRun: requireConfigName,
+	Run: func(cmd *cobra.Command, args []string) {
+		configName := re.ReplaceAllString(args[0], "")
```
I think we should also make `configName` lower case. Otherwise, you may end up with users trying and failing at:
```
$ cilium config get IPAM
Error: Configuration does not exist
```

pchaigno commented on cilium/cmd/helpers.go:
```diff
@@ -97,6 +98,16 @@ func requirePath(cmd *cobra.Command, args []string) {
 	}
 }
 
+func requireConfigName(cmd *cobra.Command, args []string) {
```
This function should be in `endpoint_config_get.go` since it appears to only be used there.

pchaigno reviewed (changes requested):
Nothing major. I think we're getting close to being able to merge this :-)

christarazi commented on cilium/cmd/config.go:
```diff
@@ -37,6 +42,7 @@ var configCmd = &cobra.Command{
 func init() {
 	rootCmd.AddCommand(configCmd)
 	configCmd.Flags().BoolVarP(&listOptions, "list-options", "", false, "List available options")
+	configCmd.Flags().BoolVarP(&listReadOnlyConfigurations, "read-only", "r", false, "Display Read Only Configurations")
```
```suggestion
	configCmd.Flags().BoolVarP(&listReadOnlyConfigurations, "read-only", "r", false, "Display read only configurations")
```

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+		} else {
+			if err := command.PrintOutput(cfgStatus.Realized); err != nil {
+				os.Exit(1)
+			}
```
I think we should be using `Fatalf` here like other Cilium commands. Plus, by just exiting, we don't present the error to the user.

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+}
+
+func dumpReadOnlyConfigs(cfgStatus *models.DaemonConfigurationStatus) {
+	fmt.Print("#### Read-Only Configurations ####\n")
```
Nit: 
```suggestion
	fmt.Println("#### Read-Only Configurations ####")
```

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+}
+
+func dumpReadWriteConfigs(cfgStatus *models.DaemonConfigurationStatus) {
+	fmt.Print("##### Read-Write Configurations #####\n")
```
```suggestion
	fmt.Println("##### Read-Write Configurations #####")
```

christarazi commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,84 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
We've switched to a shorter form of putting the license text in each file, see https://github.com/cilium/cilium/blob/master/cilium/cmd/ip.go#L1.

christarazi commented on cilium/cmd/config_get.go:
```diff
@@ -0,0 +1,84 @@
+
+	"github.com/cilium/cilium/pkg/command"
+	"github.com/cilium/cilium/test/helpers"
+	"github.com/spf13/cobra"
```
```suggestion
	"github.com/cilium/cilium/pkg/command"
	"github.com/cilium/cilium/test/helpers"
	
	"github.com/spf13/cobra"
```

christarazi commented on cilium/cmd/helpers.go:
```diff
@@ -406,3 +424,10 @@ func getIpv6EnableStatus() bool {
+		m2[k] = v
+	}
+	return m2
+}
```
This is actually modifying the 2nd argument. I think you might want to create a new map and return that instead, i.e. `m3`

christarazi reviewed (changes requested):
Well done! I have a few mostly minor comments.

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+	if command.OutputJSON() {
+		if listReadOnlyConfigurations {
+			if err := command.PrintOutput(cfgStatus.DaemonConfigurationMap); err != nil {
+				Fatalf("Cannot print json output: %v", err)
```
"Cannot print json output" doesn't really mean much to the user. What really is happening is that we cannot show the configuration:

```suggestion
				Fatalf("Cannot show configuration: %v", err)
```

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+}
+
+func dumpReadOnlyConfigs(cfgStatus *models.DaemonConfigurationStatus) {
+	fmt.Println("#### Read-Only Configurations ####")
```
```suggestion
	fmt.Println("#### Read-only configurations ####")
```

christarazi commented on cilium/cmd/config.go:
```diff
@@ -111,3 +104,62 @@ func configDaemon(cmd *cobra.Command, opts []string) {
+}
+
+func dumpReadWriteConfigs(cfgStatus *models.DaemonConfigurationStatus) {
+	fmt.Println("##### Read-Write Configurations #####")
```
```suggestion
	fmt.Println("##### Read-write configurations #####")
```

christarazi commented on Documentation/cmdref/cilium_config.md:
```diff
@@ -15,6 +15,7 @@ cilium config [<option>=(enable|disable) ...] [flags]
       --list-options    List available options
   -n, --num-pages int   Number of pages for perf ring buffer. New values have to be > 0
   -o, --output string   json| jsonpath='{}'
+  -r, --read-only       Display Read Only Configurations
```
You'll need to update the cmdref as well. `make -C Documentation update-cmdref`.

pchaigno commented on pkg/option/config.go:
```diff
@@ -1449,11 +1449,12 @@ type DaemonConfig struct {
+	EnableBPFMasquerade  bool
+	EnableBPFClockProbe  bool
+	EnableIPMasqAgent    bool
+	// expired DNS lookups with still-active connections
```
My real concern with this code change was the new comment. I understand why the other lines have changed given the new comment, but why was this comment added here? It seems unrelated to the other changes in this commit and in this PR.

h3llix replied:
Ah, eagle's eye. Sorry, I was so careless to ignore that I added a comment by mistake. I'll just fix it.
